google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1416,46 +1436,93 @@ PooledSession getReadSession() throws SpannerException {
   PooledSession getReadWriteSession() {
     Span span = Tracing.getTracer().getCurrentSpan();
     span.addAnnotation("Acquiring read write session");
-    Waiter waiter = null;
     PooledSession sess = null;
-    synchronized (lock) {
-      if (closureFuture != null) {
-        span.addAnnotation("Pool has been closed");
-        throw new IllegalStateException("Pool has been closed");
+    // Loop to retry SessionNotFoundExceptions that might occur during in-process prepare of a
+    // session.
+    while (true) {
+      Waiter waiter = null;
+      boolean inProcessPrepare = false;
+      synchronized (lock) {
+        if (closureFuture != null) {
+          span.addAnnotation("Pool has been closed");
+          throw new IllegalStateException("Pool has been closed");
+        }
+        if (resourceNotFoundException != null) {
+          span.addAnnotation("Database has been deleted");
+          throw SpannerExceptionFactory.newSpannerException(
+              ErrorCode.NOT_FOUND,
+              String.format(
+                  "The session pool has been invalidated because a previous RPC returned 'Database not found': %s",
+                  resourceNotFoundException.getMessage()),
+              resourceNotFoundException);
+        }
+        sess = writePreparedSessions.poll();
+        if (sess == null) {
+          if (numSessionsBeingPrepared <= prepareThreadPoolSize) {
+            if (numSessionsBeingPrepared <= readWriteWaiters.size()) {
+              PooledSession readSession = readSessions.poll();
+              if (readSession != null) {
+                span.addAnnotation(
+                    "Acquired read only session. Preparing for read write transaction");
+                prepareSession(readSession);
+              } else {
+                span.addAnnotation("No session available");
+                maybeCreateSession();
+              }
+            }
+          } else {
+            inProcessPrepare = true;
+            numSessionsInProcessPrepared++;
+            PooledSession readSession = readSessions.poll();
+            if (readSession != null) {
+              // Create a read/write transaction in-process if there is already a queue for prepared
+              // sessions. This is more efficient than doing it asynchronously, as it scales with
+              // the number of user threads. The thread pool for asynchronously preparing sessions
+              // is fixed.
+              span.addAnnotation(
+                  "Acquired read only session. Preparing in-process for read write transaction");
+              sess = readSession;
+            } else {
+              span.addAnnotation("No session available");
+              maybeCreateSession();
+            }
+          }
+          if (sess == null) {
+            waiter = new Waiter();
+            if (inProcessPrepare) {
+              readWaiters.add(waiter);
+            } else {
+              readWriteWaiters.add(waiter);
+            }
```

Comment:
I don't quite understand this. Why is the waiter being added to `readWaiters` if `inProcessPrepare`? Aren't we trying to create a read/write session here?

Reply:
I've added a comment on why we add it to `readWaiters` here.

The reasons are:
1. The normal prepare process is executed using a background executor with a fixed thread pool. When the work queue for this executor has exceeded the number of threads available in the thread pool, any requester for a read/write session will have to wait for a thread to come available, and then for the prepare to actually be executed. In the meantime, the user thread is just waiting. In those cases, it's more efficient to use the user thread to execute the `BeginTransaction` RPC, instead of letting it wait for a thread to come available in the thread pool.
1. In this specific case, in addition to the work queue for preparing sessions being longer than the number of threads in the thread pool, there is also no session at all available in the session pool. In that case, it's more efficient to wait for any (read) session to come available, and then execute the `BeginTransaction` RPC using the user thread, than to wait for both a session to come available, and then for a thread in the thread pool to come available for preparing the session.